File: evd_authoring_app/src/model/evd_script/program/tasks/initialize.js
import { OpenGripper } from '.';
import { NodeParser } from '../../utility_functions';
import { MachineInitialize } from '../machine_operations';
import { MoveUnplanned } from '../primitives';
import { Task } from '../task';


export class Initialize extends Task {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'initialize.';
    }

    static fullTypeString() {
        return Task.fullTypeString() + Initialize.typeString();
    }

    constructor(homeLocUuid=null, machineUuids=[], type='', name='', uuid=null, parent=null, appendType=true, primitives=null) {

        if (primitives === null) {
            primitives = [];

            primitives = primitives + machineUuids.map(id => MachineInitialize(id));

            primitives = primitives + [
                MoveUnplanned(homeLocUuid,true),
                OpenGripper()
            ];
        }
    }

    static fromDict(dct) {
        return Initialize(
            type= dct.type,
            appendType= false,
            name= dct.name,
            uuid= dct.uuid,
            primitives= dct.primitives.map(p => NodeParser(p))
        );
    }
}

File: evd_ros_backend/evd_ros_core/src/evd_script/program_nodes/hierarchical_tasks/initialize.py
'''
Initialize should be the first skill evoked in the program to
configure all machines and jog cobot to initial position.
'''

from ..hierarchical import Hierarchical
from ..machine_operations import MachineInitialize
from ..primitives import MoveUnplanned
from .open_gripper import OpenGripper


class Initialize(Hierarchical):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'initialize' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Hierarchical.full_type_string() + cls.type_string()

    def __init__(self, homeLocUuid=None, machineUuids=[], type='', name='',
                 uuid=None, parent=None, append_type=True, primitives=None,
                 editable=True, deleteable=True, description=''):

        if primitives == None:
            primitives = []

            primitives += [MachineInitialize(id, editable=editable, deleteable=False) for id in machineUuids]

            primitives += [
                MoveUnplanned(homeLocUuid,True, editable=editable, deleteable=False),
                OpenGripper(editable=editable, deleteable=False)
            ]

        super(Initialize,self).__init__(
            type=Initialize.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type,
            primitives=primitives,
            editable=editable,
            deleteable=deleteable,
            description=description)

    '''
    Execution methods
    '''

    def symbolic_execution(self, hooks):
        next = super(Initialize,self).symbolic_execution(hooks)

        # Since we are initializing state, the ambigous movement is expected and does not
        # warrant special consideration
        hooks.tokens['robot']['state']['gripper']['ambiguous_flag'] = False

        return next

    def realtime_execution(self, hooks):
        next = super(Initialize,self).realtime_execution(hooks)

        # Since we are initializing state, the ambigous movement is expected and does not
        # warrant special consideration
        hooks.tokens['robot']['state']['gripper']['ambiguous_flag'] = False

        return next
